File: Payload_Type/merlin/mythic/agent_functions/make_token.py

from merlin import *
from mythic_payloadtype_container.MythicCommandBase import *
from mythic_payloadtype_container.MythicRPC import *
import json

# Set to enable debug output to Mythic
debug = False


class MakeTokenArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
            CommandParameter(
                name="user",
                cli_name="user",
                display_name="Username",
                type=ParameterType.String,
                description="Domain and username to make a token for (e.g. ACME\\RASTLEY",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=1,
                        required=True,
                    ),
                ],
            ),
            CommandParameter(
                name="pass",
                cli_name="pass",
                display_name="Password",
                type=ParameterType.String,
                description="The account's password",
                parameter_group_info=[
                    ParameterGroupInfo(
                        group_name="Default",
                        ui_position=2,
                        required=True,
                    ),
                ],
            ),
        ]

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            if self.command_line[0] == '{':
                self.load_args_from_json_string(self.command_line)
            else:
                args = str.split(self.command_line)
                if len(args) > 0:
                    self.add_arg("user", args[0])
                if len(args) > 1:
                    self.add_arg("pass", args[1])


class MakeTokenCommand(CommandBase):
    cmd = "make_token"
    needs_admin = False
    help_cmd = "make_token <DOMAIN\\Username> <password>"
    description = "Create a new type-9 logon session and Windows access token for the provided credentials.\n" \
                  "The token is only used for NETWORK authentication, not local."
    version = 1
    author = "@Ne0nd0g"
    argument_class = MakeTokenArguments
    attackmapping = ["T1134.003"]
    attributes = CommandAttributes(
        spawn_and_injectable=False,
        supported_os=[SupportedOS.Windows]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        if debug:
            await MythicRPC().execute(function_name="create_output", task_id=task.id, output=f'\n[DEBUG]Input task:{task}')

        task.display_params = f'User: {task.args.get_arg("user")}, Password: {task.args.get_arg("pass")}'

        command = {
            "command": "token",
            "args": ["make", f'{task.args.get_arg("user")}', f'{task.args.get_arg("pass")}'],
        }

        task.args.remove_arg("user")
        task.args.remove_arg("pass")
        task.args.add_arg("type", MerlinJob.MODULE, ParameterType.Number)
        task.args.add_arg("payload", json.dumps(command), ParameterType.String)

        if debug:
            await MythicRPC().execute("create_output", task_id=task.id, output=f'\n[DEBUG]Returned task:\r\n{task}\r\n')

        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/merlin/container/commands/makeToken.go
/*
Merlin is a post-exploitation command and control framework.

This file is part of Merlin.
Copyright (C) 2023  Russel Van Tuyl

Merlin is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as
published by the Free Software Foundation, either version 3 of the License, or any later version.

Merlin is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with Merlin.  If not, see <http://www.gnu.org/licenses/>.
*/

package commands

import (
	// Standard
	"fmt"

	// Mythic
	structs "github.com/MythicMeta/MythicContainer/agent_structs"

	// Merlin Message
	"github.com/Ne0nd0g/merlin-message/jobs"
)

// makeToken creates and return a Mythic Command structure that is registered with the Mythic server
func makeToken() structs.Command {
	attr := structs.CommandAttribute{
		SupportedOS: []string{structs.SUPPORTED_OS_WINDOWS},
	}

	user := structs.CommandParameter{
		Name:                                    "user",
		ModalDisplayName:                        "Username",
		CLIName:                                 "user",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "Domain and username to make a token for (e.g. ACME\\RASTLEY",
		Choices:                                 nil,
		DefaultValue:                            nil,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       0,
				AdditionalInformation: nil,
			},
		},
	}

	pass := structs.CommandParameter{
		Name:                                    "pass",
		ModalDisplayName:                        "Password",
		CLIName:                                 "pass",
		ParameterType:                           structs.COMMAND_PARAMETER_TYPE_STRING,
		Description:                             "The account's plain-text password",
		Choices:                                 nil,
		DefaultValue:                            nil,
		SupportedAgents:                         nil,
		SupportedAgentBuildParameters:           nil,
		ChoicesAreAllCommands:                   false,
		ChoicesAreLoadedCommands:                false,
		FilterCommandChoicesByCommandAttributes: nil,
		DynamicQueryFunction:                    nil,
		ParameterGroupInformation: []structs.ParameterGroupInfo{
			{
				ParameterIsRequired:   true,
				GroupName:             "Default",
				UIModalPosition:       1,
				AdditionalInformation: nil,
			},
		},
	}

	command := structs.Command{
		Name:                  "make_token",
		NeedsAdminPermissions: false,
		HelpString:            "make_token <DOMAIN\\Username> <password>",
		Description: "Create a new type-9 logon session and Windows access token for the provided" +
			" credentials. The token is only used for NETWORK authentication, not local.",
		Version:                        0,
		SupportedUIFeatures:            nil,
		Author:                         "@Ne0nd0g",
		MitreAttackMappings:            []string{"T1134.003"},
		ScriptOnlyCommand:              false,
		CommandAttributes:              attr,
		CommandParameters:              []structs.CommandParameter{user, pass},
		AssociatedBrowserScript:        nil,
		TaskFunctionOPSECPre:           nil,
		TaskFunctionCreateTasking:      makeTokenCreateTask,
		TaskFunctionProcessResponse:    nil,
		TaskFunctionOPSECPost:          nil,
		TaskFunctionParseArgString:     taskFunctionParseArgString,
		TaskFunctionParseArgDictionary: taskFunctionParseArgDictionary,
		TaskCompletionFunctions:        nil,
	}

	return command
}

// makeTokenCreateTask takes a Mythic Task and converts into a Merlin Job that is encoded into JSON and subsequently sent to the Merlin Agent
func makeTokenCreateTask(task *structs.PTTaskMessageAllData) (resp structs.PTTaskCreateTaskingMessageResponse) {
	resp.TaskID = task.Task.ID

	user, err := task.Args.GetStringArg("user")
	if err != nil {
		resp.Error = fmt.Sprintf("mythic/container/commands/makeToken/makeTokenCreateTask(): %s", err)
		resp.Success = false
		return
	}

	pass, err := task.Args.GetStringArg("pass")
	if err != nil {
		resp.Error = fmt.Sprintf("mythic/container/commands/makeToken/makeTokenCreateTask(): %s", err)
		resp.Success = false
		return
	}

	job := jobs.Command{
		Command: "token",
		Args:    []string{"make", user, pass},
	}

	mythicJob, err := ConvertMerlinJobToMythicTask(job, jobs.MODULE)
	if err != nil {
		resp.Error = fmt.Sprintf("mythic/container/commands/makeToken/makeTokenCreateTask(): %s", err)
		resp.Success = false
		return
	}

	task.Args.SetManualArgs(mythicJob)

	disp := fmt.Sprintf("%s %s", user, pass)
	resp.DisplayParams = &disp
	resp.Success = true

	return
}
